### PR TITLE
[qt] Move some WalletModel functions into CWallet

### PR DESCRIPTION
**Pull request by ryanofsky**
Motivation for moving these is to make supporting IPC simpler (#10102), so these lookups can be one-shot IPC requests, instead of back-and-forth interactions over the IPC channel.

Also these functions are potentially useful outside of the bitcoin GUI (e.g. for RPCs).

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Yes. They belong to CWallet rather then WalletModel. The coincontrol'w `ListCoins` function can potentially be useful for non GUI features. My only nit is if we should rename `ListCoins` to something more meaningful.

Concept ACK.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2060,6 +2080,59 @@ void CWallet::AvailableCoins(std::vector<COutput>& vCoins, bool fOnlySafe, const
...
+    for (const auto& output : lockedCoins) {
+        auto it = mapWallet.find(output.hash);
+        if (it != mapWallet.end()) {
+            int depth = it->second.GetDepthInMainChain();
+            if (depth >= 0 && output.n < it->second.tx->vout.size() &&
+                IsMine(it->second.tx->vout[output.n]) == ISMINE_SPENDABLE) {
+                CTxDestination address;
+                if (ExtractDestination(FindNonChangeParentOutput(*it->second.tx, output.n).scriptPubKey, address)) {
+                    result[address].emplace_back(
+                        &it->second, output.n, depth, true /* spendable */, true /* solvable */, false /* safe */);
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: Why did you change to false for safe here...looks like the old code did true? Does that have any effect?

File: src/qt/walletmodel.cpp [hunk 2]
```diff
@@ -596,37 +589,11 @@ bool WalletModel::isSpent(const COutPoint& outpoint) const
 void WalletModel::listCoins(std::map<QString, std::vector<COutput> >& mapCoins) const
 {
     std::vector<COutput> vCoins;
```
Comment (TheBlueMatt) [inline src/qt/walletmodel.cpp]: We dont need vCoins anymore either, no?

**Review by TheBlueMatt — COMMENTED** [review]
Looks good, seems reasonable, though I hope we can delete half the code here (see #10337).

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -1983,6 +1990,19 @@ CAmount CWallet::GetLegacyBalance(const isminefilter& filter, int minDepth, cons
     return balance;
 }
 
+CAmount CWallet::GetAvailableBalance(const CCoinControl* coinControl) const
+{
+    CAmount balance = 0;
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Not sure if a lock is required at this point. Do you want to protect `const CCoinControl* coinControl`?
`AvailableCoins()` does protect via `cs_main`/`cs_wallet`.

Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: We have pointers in the COutput to entires in mapWallet, which can (theoretically) be delted out from under us thanks to the zap rpc.

**Review by ryanofsky — COMMENTED** [review]
Squashed 14c1174c9aadb446c9a9e16626611f476610f4ba -> b6a9cc48fb950d1f0daa6c3db1df2bbc2b591ab0 ([pr/ipc-move.2](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.2) -> [pr/ipc-move.3](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.3))

File: src/qt/test/wallettests.cpp [hunk 4]
```diff
@@ -80,8 +79,9 @@ QModelIndex FindTx(const QAbstractItemModel& model, const uint256& txid)
 //     src/qt/test/test_bitcoin-qt -platform xcb      # Linux
 //     src/qt/test/test_bitcoin-qt -platform windows  # Windows
 //     src/qt/test/test_bitcoin-qt -platform cocoa    # macOS
-void WalletTests::walletTests()
+void TestSendCoins()
 {
+    return;
```
Comment (TheBlueMatt) [inline src/qt/test/wallettests.cpp]: Were you planning on re-enabling this?

Comment (ryanofsky) [inline src/qt/test/wallettests.cpp]: Wow, good catch. Fixed in 01f847bf2b0ea395f8533e4d52c9422807af0566. This was just an accident, I didn't mean to disable it at all.

**Review by ryanofsky — COMMENTED** [review]
Squashed 01f847bf2b0ea395f8533e4d52c9422807af0566 -> 38d3864eccd6abbbcf22619c8109a1b6e015f901 ([pr/ipc-move.4](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.4) -> [pr/ipc-move.5](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.5))

File: src/qt/walletmodel.cpp [hunk 5]
```diff
@@ -657,10 +623,7 @@ void WalletModel::listLockedCoins(std::vector<COutPoint>& vOutpts)
 void WalletModel::loadReceiveRequests(std::vector<std::string>& vReceiveRequests)
 {
     LOCK(wallet->cs_wallet);
```
Comment (TheBlueMatt) [inline src/qt/walletmodel.cpp]: nit: maybe push this lock down into GetDestValues?

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -2060,6 +2082,59 @@ void CWallet::AvailableCoins(std::vector<COutput>& vCoins, bool fOnlySafe, const
     }
 }
 
+std::map<CTxDestination, std::vector<COutput>> CWallet::ListCoins() const
+{
+    std::map<CTxDestination, std::vector<COutput>> result;
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: nit: might be nice to AssertLockHeld here (because you cant call this without cs_wallet due to the COutput returns).

**Review by TheBlueMatt — COMMENTED** [review]
utACK 38d3864eccd6abbbcf22619c8109a1b6e015f901

**Review by ryanofsky — COMMENTED** [review]
Squashed fe8c036cb87c62bee653fe581544c714ce8808d2 -> a226c97c336c44b4b8320328debea69e576f0fa8 ([pr/ipc-move.6](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.6) -> [pr/ipc-move.7](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.7))

**Comment by ryanofsky** [comment]
Rebased a226c97c336c44b4b8320328debea69e576f0fa8 -> 108f04f2d973adac5313c7e4e17a59766a3cc1b6 ([pr/ipc-move.7](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.7) -> [pr/ipc-move.8](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-move.8)) because of conflict with #8952.

**Comment by laanwj** [comment]
utACK 108f04f
